projects/lib/src/board/gryphonboard.cpp

```diff
@@ -0,0 +1,315 @@
+		  return;
+
+	int srcType = piece.type();
+	int tgtType = successorType(srcType, false);
+
+	if (srcType != tgtType)
+	{
+		piece.setType(tgtType);
+		setSquare(target, piece);
+	}
+}
+
+void GryphonBoard::vUndoMove(const Move& move)
+{
+	int target = move.targetSquare();
+	Piece piece = pieceAt(target);
+	Piece oldpiece = m_pieceStack.pop();
+	Q_ASSERT(oldpiece.side() == piece.side());
+	piece.setType(oldpiece.type());
+	setSquare(target, piece);
+
+	WesternBoard::vUndoMove(move);
+}
+
+bool GryphonBoard::isLegalPosition()
+{
+	if (!WesternBoard::isLegalPosition())
+		return false;
+
+	int count[2][King+1] = {{0, 15, 4, 4, 4, 2, 15},
+				{0, 15, 4, 4, 4, 2, 15}};
+
+	for (int sq = m_start; sq < m_end; sq++)
+	{
+		Piece piece = pieceAt(sq);
+		if (piece.isValid())
+			if (--count[piece.side()][piece.type()] < 0)
+				// too many pieces of this type
+				return false;
+	}
+
+	return true;
+}
+
+void GryphonBoard::generateMovesForPiece(QVarLengthArray< Move >& moves,
+					 int pieceType,
+					 int square) const
+{
+	// return early if successor of pieceType is already fully occupied
+	Side side = sideToMove();
+	int successor = successorType(pieceType);
+	static const int count[King+1] = {0, 15, 4, 4, 4, 2, 15};
+	int cnt = count[successor];
+
+	for (int sq = m_start; sq < m_end; sq++)
+	{
+		if (pieceAt(sq) == Piece(side, successor))
+			if (--cnt < 1)
+				return;
+	}
+
+	WesternBoard::generateMovesForPiece(moves, pieceType, square);
+}
+
+
+
+SimplifiedGryphonBoard::SimplifiedGryphonBoard()
+	: GryphonBoard(),
+	  m_captures {0, 0}
+{
+}
+
+Board* SimplifiedGryphonBoard::copy() const
+{
+	return new SimplifiedGryphonBoard(*this);
+}
+
+QString SimplifiedGryphonBoard::variant() const
+{
+	return "simplifiedgryphon";
+}
+
+QString SimplifiedGryphonBoard::defaultFenString() const
+{
+	return "4k3/pppppppp/8/8/8/8/PPPPPPPP/4K3 w - - 0 1";
+}
+
+void SimplifiedGryphonBoard::generateMovesForPiece(QVarLengthArray< Move >& moves,
+						   int pieceType,
+						   int square) const
+{
+	if (pieceType != King)
+		return GryphonBoard::generateMovesForPiece(moves, pieceType, square);
+
+	QVarLengthArray< Move > newmoves;
+	GryphonBoard::generateMovesForPiece(newmoves, King, square);
+
+	for (const Move move: newmoves)
+	{
+		Square srcSq = chessSquare(move.sourceSquare());
+		Square tgtSq = chessSquare(move.targetSquare());
+		int rrank = tgtSq.rank() - srcSq.rank();
+
+		if (sideToMove() == Side::Black)
+			rrank = -rrank;
+
+		// side without captures must not move king backward or sideways
+		if (m_captures[sideToMove()] > 0 || rrank > 0)
+			moves.append(move);
+	}
+}
+
+void SimplifiedGryphonBoard::vMakeMove(const Move& move,
+				       BoardTransition* transition)
+{
+	if(captureType(move) != Piece::NoPiece)
+		m_captures[sideToMove()]++;
+
+	GryphonBoard::vMakeMove(move, transition);
+}
+
+void SimplifiedGryphonBoard::vUndoMove(const Move& move)
+{
+	GryphonBoard::vUndoMove(move);
+
+	if(captureType(move) != Piece::NoPiece)
```

Comment:
We put a space between `if` and parenthesis.

Reply:
+1